src/checkers/cwe_476.ml

```diff
@@ -1,47 +1,280 @@
+        flag_all_unchecked_registers state cwe_hits
+      | false -> (* we assume that the callee will check all remaining unchecked values *)
+        [] in
+    match Call.target call with
+    | Indirect(_) -> state (* already handled above *)
+    | Direct(tid) ->
+      if List.exists function_names ~f:(fun elem -> String.(=) elem (Tid.name tid)) then
+        let func = Term.find_exn sub_t program tid in
+        let arguments = Term.enum arg_t func in
+        (* Every return register is tainted as unchecked return value. *)
+        Seq.fold arguments ~init:state ~f:(fun state arg ->
+            match Bap.Std.Arg.intent arg with
+            | None | Some(In) -> state
+            | Some(Out) | Some(Both) ->
+              let variable = match Bap.Std.Arg.rhs arg with
+                | Bil.Var(var) -> var
+                | _ -> failwith "[CWE476] Return register wasn't a register." in
+              State.add state variable (Term.tid block)
+          )
+      else
+        state
+
+(** updates a block analysis.
+    The strict call policy decides the behaviour on call and return instructions:
+    strict: all unchecked values get flagged as cwe-hits
+    non-strict: the state gets cleared, it is assumed that the target of the call/return
+    instruction checks all remaining unchecked values. *)
+let update_block_analysis block register_state ~cwe_hits ~function_names ~program ~strict_call_policy =
+  let elements = Blk.elts block in
+  let register_state = Seq.fold elements ~init:register_state ~f:(fun state element ->
+      match element with
+      | `Def def -> update_state_def def state ~cwe_hits
+      | `Phi phi -> state (* We ignore phi terms for this analysis. *)
+      | `Jmp jmp -> update_state_jmp jmp state ~cwe_hits ~function_names ~program ~block ~strict_call_policy
+    ) in
+  State.remove_virtual_registers register_state (* virtual registers should not be accessed outside of the block where they are defined. *)
+
+let print_hit tid ~sub ~function_names ~tid_map =
+  let block = Option.value_exn (Term.find blk_t sub tid) in
+  let jmps = Term.enum jmp_t block in
+  let _ = Seq.find_exn jmps ~f:(fun jmp ->
+      match Jmp.kind jmp with
+      | Call(call) -> begin
+          match Call.target call with
+          | Direct(call_tid) -> Option.is_some (List.find function_names ~f:(fun name ->
+              if name = (Tid.name call_tid) then begin
+                Log_utils.warn "[%s] {%s} (NULL Pointer Dereference) There is no check if the return value is NULL at %s (%s)."
+                  name
+                  version
+                  (Address_translation.translate_tid_to_assembler_address_string tid tid_map)
+                  name;
+                true
+              end else
+                false
+            ))
+          | _ -> false
+        end
+      | _ -> false
+    ) in ()
+
+let check_cwe prog proj tid_map symbol_pairs =
+  let strict_call_policy = true in
```

Comment:
make this configurable in JSON config since it might be a good way to tune false positives/false negatives. Document this parameter accordingly.

Reply:
I've added a general way to add parameters to the checks configurable in config.json.